### PR TITLE
Feat/oidf 51

### PR DESCRIPTION
**Pull request by jcmelati**

### CODE REVIEW

File: modules/admin-server/src/main/kotlin/com/sphereon/oid/fed/server/admin/controllers/AccountController.kt [hunk 1]
```diff
@@ -0,0 +1,22 @@
...
+    @GetMapping
+    fun getAccounts(): List<AccountDTO> {
+        return accountService.findAll()
+    }
+
+    @PostMapping
+    fun createAccount(@RequestBody account: CreateAccountDTO) {
+        return accountService.create(account)
+    }
+}
```
Comment (maikel-maas) [inline modules/admin-server/src/main/kotlin/com/sphereon/oid/fed/server/admin/controllers/AccountController.kt:22]: Add trailing new line

File: modules/admin-server/src/main/kotlin/com/sphereon/oid/fed/server/admin/services/AccountService.kt [hunk 2]
```diff
@@ -0,0 +1,18 @@
...
+    private val accountRepository = Persistence.accountRepository
+
+    fun create(account: CreateAccountDTO) {
+        return accountRepository.create(account)
+    }
+
+    fun findAll(): List<AccountDTO> {
+        return accountRepository.findAll().map { it.toDTO() }
+    }
+}
```
Comment (maikel-maas) [inline modules/admin-server/src/main/kotlin/com/sphereon/oid/fed/server/admin/services/AccountService.kt:18]: Add trailing new line

File: modules/persistence/build.gradle.kts [hunk 3]
```diff
@@ -0,0 +1,51 @@
...
+            }
+        }
+    }
+}
+
+java {
+    toolchain {
+        languageVersion = JavaLanguageVersion.of(21)
+    }
+}
```
Comment (maikel-maas) [inline modules/persistence/build.gradle.kts:51]: Add trailing new line

File: modules/persistence/build.gradle.kts [hunk 4]
```diff
@@ -0,0 +1,51 @@
...
+                implementation("org.postgresql:postgresql:42.7.3")
+                implementation("app.cash.sqldelight:jdbc-driver:2.0.2")
+            }
+        }
+    }
+}
+
+java {
+    toolchain {
+        languageVersion = JavaLanguageVersion.of(21)
```
Comment (maikel-maas) [inline modules/persistence/build.gradle.kts:49]: Will not work for multiplatform. It should be done in the jvm target. Ex:
`kotlin {
    jvm {
        tasks {
            named<KotlinJvmCompile>("compileKotlinJvm") {
                dependsOn("fixOpenApiGeneratorIssue")
                compilerOptions {
                    jvmTarget.set(JvmTarget.JVM_11)
                }
            }
        }
    }
}`

File: modules/persistence/build.gradle.kts [hunk 5]
```diff
@@ -0,0 +1,51 @@
...
+            dialect("app.cash.sqldelight:postgresql-dialect:2.0.2")
+            packageName.set("com.sphereon.oid.fed.persistence")
+            srcDirs.from(project.file("./src/commonMain/sqldelight/"))
+        }
+    }
+}
+
+kotlin {
+    jvm{
+        withJava()
```
Comment (maikel-maas) [inline modules/persistence/build.gradle.kts:27]: Do we want to mix Java code with Kotlin?

File: modules/persistence/gradle/wrapper/gradle-wrapper.properties [hunk 6]
```diff
@@ -0,0 +1,6 @@
+#Thu Jul 25 12:58:25 CEST 2024
+distributionBase=GRADLE_USER_HOME
+distributionPath=wrapper/dists
```
Comment (maikel-maas) [inline modules/persistence/gradle/wrapper/gradle-wrapper.properties:3]: Use Unix carriage return. It should be \n and not \r\n.

File: modules/persistence/src/commonMain/kotlin/com.sphereon.oid.fed.persistence/Persistence.kt [hunk 7]
```diff
@@ -0,0 +1,7 @@
+package com.sphereon.oid.fed.persistence
+
+import com.sphereon.oid.fed.persistence.repositories.AccountRepository
+
+expect object Persistence {
+    val accountRepository: AccountRepository
+}
```
Comment (maikel-maas) [inline modules/persistence/src/commonMain/kotlin/com.sphereon.oid.fed.persistence/Persistence.kt:7]: Add trailing new line

File: modules/persistence/src/commonMain/kotlin/com.sphereon.oid.fed.persistence/database/DatabaseFactory.kt [hunk 8]
```diff
@@ -0,0 +1,21 @@
...
+            is DatabaseConfig.Sqlite -> platformSqlDriver.createSqliteDriver(config.path)
+            else -> throw IllegalArgumentException("Unsupported database config")
+        }
+    }
+}
+
+sealed class DatabaseConfig {
+    data class Postgres(val url: String, val username: String, val password: String) : DatabaseConfig()
+    data class Sqlite(val path: String) : DatabaseConfig()
+}
```
Comment (maikel-maas) [inline modules/persistence/src/commonMain/kotlin/com.sphereon.oid.fed.persistence/database/DatabaseFactory.kt:21]: Add trailing new line

File: modules/persistence/src/commonMain/kotlin/com.sphereon.oid.fed.persistence/database/DatabaseFactory.kt [hunk 9]
```diff
@@ -0,0 +1,21 @@
...
+    fun createPostgresDriver(url: String, username: String, password: String): SqlDriver
+    fun createSqliteDriver(path: String): SqlDriver
+}
+
+class DatabaseFactory(private val platformSqlDriver: PlatformSqlDriver) {
+    fun createDriver(config: DatabaseConfig): SqlDriver {
+        return when (config) {
+            is DatabaseConfig.Postgres -> platformSqlDriver.createPostgresDriver(config.url, config.username, config.password)
+            is DatabaseConfig.Sqlite -> platformSqlDriver.createSqliteDriver(config.path)
+            else -> throw IllegalArgumentException("Unsupported database config")
```
Comment (maikel-maas) [inline modules/persistence/src/commonMain/kotlin/com.sphereon.oid.fed.persistence/database/DatabaseFactory.kt:13]: Move the exception message to a constant in a specific
file

File: modules/persistence/src/commonMain/sqldelight/entities/Account.sq [hunk 10]
```diff
@@ -0,0 +1,29 @@
...
+UPDATE account SET deleted_at = CURRENT_TIMESTAMP WHERE id = ?;
+
+findByUsername:
+SELECT * FROM account WHERE username = ?;
+
+findById:
+SELECT * FROM account WHERE id = ?;
+
+update:
+UPDATE account SET username = ? WHERE id = ?;
```
Comment (maikel-maas) [inline modules/persistence/src/commonMain/sqldelight/entities/Account.sq:29]: Add trailing new line

File: modules/persistence/src/jvmMain/kotlin/com.sphereon.oid.fed.persistence/Persistence.jvm.kt [hunk 11]
```diff
@@ -0,0 +1,13 @@
...
+
+actual object Persistence {
+    actual val accountRepository: AccountRepository
+
+    init {
+        val driver = PlatformSqlDriver().createPostgresDriver(System.getenv("DB_URL"), System.getenv("DB_USER"), System.getenv("DB_PASSWORD"))
+        val database = Database(driver)
+        accountRepository = AccountRepository(database)
+    }
+}
```
Comment (maikel-maas) [inline modules/persistence/src/jvmMain/kotlin/com.sphereon.oid.fed.persistence/Persistence.jvm.kt:13]: Add trailing new line

File: modules/persistence/src/jvmMain/kotlin/com.sphereon.oid.fed.persistence/Persistence.jvm.kt [hunk 12]
```diff
@@ -0,0 +1,13 @@
+package com.sphereon.oid.fed.persistence
+import PlatformSqlDriver
+import com.sphereon.oid.fed.persistence.repositories.AccountRepository
+
+actual object Persistence {
+    actual val accountRepository: AccountRepository
+
+    init {
+        val driver = PlatformSqlDriver().createPostgresDriver(System.getenv("DB_URL"), System.getenv("DB_USER"), System.getenv("DB_PASSWORD"))
```
Comment (maikel-maas) [inline modules/persistence/src/jvmMain/kotlin/com.sphereon.oid.fed.persistence/Persistence.jvm.kt:9]: Extract a constant from strings and move them to a specific file

File: modules/persistence/src/jvmMain/kotlin/com.sphereon.oid.fed.persistence/PlatformSqlDriver.kt [hunk 13]
```diff
@@ -0,0 +1,20 @@
...
+        config.password = password
+
+        val dataSource = HikariDataSource(config)
+        return dataSource.asJdbcDriver()
+    }
+
+    actual fun createSqliteDriver(path: String): SqlDriver {
+        throw UnsupportedOperationException("Sqlite is not supported on JVM")
+    }
+}
```
Comment (maikel-maas) [inline modules/persistence/src/jvmMain/kotlin/com.sphereon.oid.fed.persistence/PlatformSqlDriver.kt:20]: Add trailing new line

File: modules/persistence/src/jvmMain/kotlin/com.sphereon.oid.fed.persistence/PlatformSqlDriver.kt [hunk 14]
```diff
@@ -0,0 +1,20 @@
...
+        config.jdbcUrl = url
+        config.username = username
+        config.password = password
+
+        val dataSource = HikariDataSource(config)
+        return dataSource.asJdbcDriver()
+    }
+
+    actual fun createSqliteDriver(path: String): SqlDriver {
+        throw UnsupportedOperationException("Sqlite is not supported on JVM")
```
Comment (maikel-maas) [inline modules/persistence/src/jvmMain/kotlin/com.sphereon.oid.fed.persistence/PlatformSqlDriver.kt:18]: Extract a constant from the exception message and move it a specific file

**Review by maikel-maas — CHANGES_REQUESTED** [review]
This implementation is creating tables, something that is already implemented using Flyway. Will Flyway be removed from the project?

File: modules/persistence/src/commonMain/sqldelight/entities/Account.sq [hunk 15]
```diff
@@ -0,0 +1,29 @@
+CREATE TABLE account (
```
Comment (maikel-maas) [inline modules/persistence/src/commonMain/sqldelight/entities/Account.sq:1]: Is table creation sqldelight responsibility?

File: modules/persistence/build.gradle.kts (hunk 4 above)
Comment (jcmelati) [reply modules/persistence/build.gradle.kts:49]: would this suffice?
`    @OptIn(ExperimentalKotlinGradlePluginApi::class)
    compilerOptions {
        apiVersion.set(org.jetbrains.kotlin.gradle.dsl.KotlinVersion.KOTLIN_2_1)
    }
`

File: modules/persistence/src/commonMain/sqldelight/entities/Account.sq (hunk 15 above)
Comment (jcmelati) [reply modules/persistence/src/commonMain/sqldelight/entities/Account.sq]: I followed the "Fresh Schema" guide from here https://cashapp.github.io/sqldelight/2.0.2/jvm_postgresql/

Comment (nklomp) [reply modules/persistence/src/commonMain/sqldelight/entities/Account.sq]: Not sure if you can combine them both, but we should follow the migration approach, ensuring a consistent DB over time

**Pull request closed without merging**